File: labday/app/src/main/java/com/jakdor/labday/view/ui/SplashFragment.java
package com.jakdor.labday.view.ui;

import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.view.animation.LinearOutSlowInInterpolator;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.ImageView;

import com.jakdor.labday.R;
import com.jakdor.labday.common.model.AppData;
import com.jakdor.labday.di.InjectableFragment;
import com.jakdor.labday.rx.RxResponse;
import com.jakdor.labday.rx.RxStatus;
import com.jakdor.labday.viewmodel.SplashViewModel;

import javax.inject.Inject;
import butterknife.BindView;
import butterknife.ButterKnife;
import timber.log.Timber;

/**
 * Loaded in MainActivity prepares {@link com.jakdor.labday.common.repository.ProjectRepository}
 * while displaying welcome logo/animations, provides seamless transition between
 * {@link SplashActivity} and {@link MainActivity}
 * */
public class SplashFragment extends Fragment implements InjectableFragment {

    @BindView(R.id.splash_logo)
    ImageView splashLogo;

    @Inject
    ViewModelProvider.Factory viewModelFactory;

    private SplashViewModel viewModel;

    private Handler delayedTransactionHandler = new Handler();
    private Runnable runnable = this::switchToLoginFragment;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = getLayoutInflater().inflate(R.layout.splash, container, false);
        ButterKnife.bind(this, view);
        startAnimations();
        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        if(viewModel == null) {
            viewModel = ViewModelProviders.of(this, viewModelFactory)
                    .get(SplashViewModel.class);
        }

        if(viewModel.isLoggedIn(getContext())) {
            observeLastUpdate();
            viewModel.updateAppData(getContext());
        }
        else {
            Timber.i("No access token, switching to loginFragment");
            delayedTransactionHandler.postDelayed(runnable, 1000);
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        delayedTransactionHandler.removeCallbacks(runnable);
    }

    public void startAnimations(){
        Animation animation = AnimationUtils.loadAnimation(getContext(), R.anim.splash_anim);
        animation.setInterpolator(new LinearOutSlowInInterpolator());
        animation.start();

        splashLogo.clearAnimation();
        splashLogo.startAnimation(animation);
    }

    public void observeLastUpdate(){
        viewModel.getResponse().observe(this, this::switchToMainFragment);
    }

    public void switchToMainFragment(RxResponse<AppData> response){
        if(response.status == RxStatus.SUCCESS || response.status == RxStatus.SUCCESS_DB) {
            MainFragment mainFragment = new MainFragment();

            if(getActivity() == null){
                Timber.wtf("Unable to get Activity");
                return;
            }

            FragmentManager fragmentManager = getActivity().getSupportFragmentManager();
            fragmentManager.popBackStack();
            fragmentManager.beginTransaction()
                    .setCustomAnimations(R.anim.fragment_fade_in_slow, R.anim.fragment_fade_out)
                    .addToBackStack(null)
                    .replace(R.id.fragmentLayout, mainFragment)
                    .commit();
        }
        else {
            Timber.wtf("Unable to get repo data, fallback to login");
            switchToLoginFragment();
        }
    }

    public void switchToLoginFragment(){
        LoginFragment loginFragment = new LoginFragment();

        if(getActivity() == null){
            Timber.wtf("Unable to get Activity");
            return;
        }

        FragmentManager fragmentManager = getActivity().getSupportFragmentManager();
        fragmentManager.popBackStack();
        fragmentManager.beginTransaction()
                .setCustomAnimations(R.anim.fragment_fade_in, R.anim.fragment_fade_out)
                .addToBackStack(null)
                .replace(R.id.fragmentLayout, loginFragment)
                .commit();
    }

    public void setViewModel(SplashViewModel viewModel) {
        this.viewModel = viewModel;
    }

    public Handler getDelayedTransactionHandler() {
        return delayedTransactionHandler;
    }
}

File: labday/app/src/main/java/com/jakdor/labday/view/ui/SplashFragment.kt
package com.jakdor.labday.view.ui

import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import android.os.Bundle
import android.os.Handler
import androidx.fragment.app.Fragment
import androidx.interpolator.view.animation.LinearOutSlowInInterpolator
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils

import com.jakdor.labday.R
import com.jakdor.labday.common.model.AppData
import com.jakdor.labday.di.InjectableFragment
import com.jakdor.labday.rx.RxResponse
import com.jakdor.labday.rx.RxStatus
import com.jakdor.labday.viewmodel.SplashViewModel
import kotlinx.android.synthetic.main.splash.*

import javax.inject.Inject
import timber.log.Timber

/**
 * Loaded in MainActivity prepares [com.jakdor.labday.common.repository.ProjectRepository]
 * while displaying welcome logo/animations, provides seamless transition between
 * [SplashActivity] and [MainActivity]
 */
class SplashFragment : Fragment(), InjectableFragment {

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    private var viewModel: SplashViewModel? = null

    val delayedTransactionHandler = Handler()
    private val runnable = Runnable { this.switchToLoginFragment() }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return layoutInflater.inflate(R.layout.splash, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        startAnimations()
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        if (viewModel == null) {
            viewModel = ViewModelProviders.of(this, viewModelFactory)
                    .get(SplashViewModel::class.java)
        }

        if (viewModel!!.isLoggedIn(context)) {
            observeLastUpdate()
            viewModel!!.updateAppData(context)
        } else {
            Timber.i("No access token, switching to loginFragment")
            delayedTransactionHandler.postDelayed(runnable, 1000)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        delayedTransactionHandler.removeCallbacks(runnable)
    }

    fun startAnimations() {
        val animation = AnimationUtils.loadAnimation(context, R.anim.splash_anim)
        animation.interpolator = LinearOutSlowInInterpolator()
        animation.start()

        splash_logo.clearAnimation()
        splash_logo.startAnimation(animation)
    }

    fun observeLastUpdate() {
        viewModel!!.response.observe(this, androidx.lifecycle.Observer { this.switchToMainFragment(it) })
    }

    fun switchToMainFragment(response: RxResponse<AppData>) {
        if (response.status == RxStatus.SUCCESS || response.status == RxStatus.SUCCESS_DB) {
            val mainFragment = MainFragment()

            if (activity == null) {
                Timber.wtf("Unable to get Activity")
                return
            }

            val fragmentManager = activity!!.supportFragmentManager
            fragmentManager.popBackStack()
            fragmentManager.beginTransaction()
                    .setCustomAnimations(R.anim.fragment_fade_in_slow, R.anim.fragment_fade_out)
                    .addToBackStack(null)
                    .replace(R.id.fragmentLayout, mainFragment)
                    .commit()
        } else {
            Timber.wtf("Unable to get repo data, fallback to login")
            switchToLoginFragment()
        }
    }

    fun switchToLoginFragment() {
        val loginFragment = LoginFragment()

        if (activity == null) {
            Timber.wtf("Unable to get Activity")
            return
        }

        val fragmentManager = activity!!.supportFragmentManager
        fragmentManager.popBackStack()
        fragmentManager.beginTransaction()
                .setCustomAnimations(R.anim.fragment_fade_in, R.anim.fragment_fade_out)
                .addToBackStack(null)
                .replace(R.id.fragmentLayout, loginFragment)
                .commit()
    }

    fun setViewModel(viewModel: SplashViewModel) {
        this.viewModel = viewModel
    }
}
